### PR TITLE
A few fixes for `pulumi logs`

### LINKED ISSUE #1828: `pulumi logs` needs to support first-class providers
Rather than using the stack's ambient configuration, `pulumi logs` implementations should use the configuration for a resource's provider.

### LINKED ISSUE #1926: `pulumi logs` doesn't emit an error when logs are unreachable?
I have a set of applications that I deploy to different AWS accounts, and I use the [assume-role](https://github.com/remind101/assume-role) tool to run `pulumi up` under an assumed role when I do a deployment.

However, I often forget to run `assume-role` before doing a `pulumi logs` command.  Instead of failing, however, it just prints that there were no logs, which always confuses me.  Then I remember I need to run `pulumi logs` after assuming the role and I'm able to look at logs.

It might be nice if the tool mentioned that logs from some sources couldn't be collected?

### LINKED ISSUE #2665: Regression? `pulumi logs` not working for `@pulumi/cloud` functions
I'm experiencing this issue that was apparently fixed in November: https://github.com/pulumi/pulumi/issues/2156

Logs from lambdas created with `aws.lambda.createFunctionFromEventHandler` work fine, but if I use `cloud.createFunction`, then I get no log output.

### LINKED ISSUE #3947: Error with `pulumi logs -f` in Go Pulumi app
`pulumi logs -f` fails with the error `no AWS region found` for each resource in my Pulumi application when my Pulumi application is a **_Go_** application. I've reproduced this across two distinct (but similar) Go Pulumi applications.

Pulumi version is `v1.10.1`.

#### Steps to reproduce

1. `pulumi up` on the [aws-go-lambda](https://github.com/pulumi/examples/tree/master/aws-go-lambda) example 
1. `pulumi logs -f` - observe error

#### Output

```
aws-go-lambda % pulumi up -y --skip-preview
Updating (dev):

     Type                    Name               Status
 +   pulumi:pulumi:Stack     go-lambda-dev      created
 +   ├─ aws:iam:Role         task-exec-role     created
 +   ├─ aws:iam:RolePolicy   lambda-log-policy  created
 +   └─ aws:lambda:Function  basicLambda        created

Outputs:
    lambda: "arn:aws:lambda:us-west-2:052848974346:function:basicLambda-cf098f0"

Resources:
    + 4 created

Duration: 38s

Permalink: https://app.pulumi.com/clstokes/go-lambda/dev/updates/2
aws-go-lambda % pulumi logs -f
Collecting logs for stack dev since 2020-02-20T10:56:46.000-08:00.

error: failed to get logs: 3 errors occurred:
	* no AWS region found
	* no AWS region found
	* no AWS region found

```

### PR DESCRIPTION
**Pull request by lukehoban**
While many of the most fundamental limitations of `pulumi logs` are still dependent on moving this code out of pulumi/pulumi and into pulumi/pulumi-aws and other resource providers (https://github.com/pulumi/pulumi/issues/608), this PR implements a few more tactical fixes which improve some common scenarios.

Unfortunately, this whole area of the code is largely untested, in part because we didn't want to add a dependency on AWS in this layer (yet another reason we really want to do https://github.com/pulumi/pulumi/issues/608).  We could choose to further violate this layering and add tests here that use `pulumi-aws` to deploy Pulumi resources and then test them against this AWS operations provider.

In the meantime, I have manually validated fixes for the following issues.
    
Fixes #2665.
Fixes #1926.
Fixes #3947.
Fixes #1828.

### CODE REVIEW

File: pkg/operations/operations_aws.go [hunk 1]
```diff
@@ -71,6 +84,14 @@ func AWSOperationsProvider(
 	return prov, nil
 }
 
+func getPropertyMapStringValue(m resource.PropertyMap, k resource.PropertyKey) string {
+	v, ok := m[k]
+	if !ok {
+		return ""
+	}
+	return v.StringValue()
```
Comment (Frassle) [inline pkg/operations/operations_aws.go:92]: StringValue panics if it isn't a string. This should probably check IsString first.

File: pkg/operations/operations_aws.go [hunk 2]
```diff
@@ -130,33 +152,48 @@ type awsConnection struct {
 }
 
 var (
-	awsDefaultSession      *session.Session
+	awsDefaultSessions     map[string]*session.Session = map[string]*session.Session{}
 	awsDefaultSessionMutex sync.Mutex
 )
 
-func getAWSSession(awsRegion, awsAccessKey, awsSecretKey, token string) (*session.Session, error) {
+func getAWSSession(awsRegion, awsAccessKey, awsSecretKey, awsToken, awsProfile string, testSession bool) (*session.Session, error) {
```
Comment (Frassle) [inline pkg/operations/operations_aws.go:159]: Most of the args are obvious by name, but a comment for what testSession is for would make this clearer. And maybe validate would be a better name for it?

File: changelog/pending/20230727--cli--fixes-for-pulumi-import.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: fix
+  scope: cli
+  description: Several fixes for `pulumi import` including support for first-class providers, support for ambient credentials and improved error reporting.
```
Comment (Frassle) [inline changelog/pending/20230727--cli--fixes-for-pulumi-import.yaml:4]:
```suggestion
  description: Several fixes for `pulumi logs` including support for first-class providers, support for ambient credentials and improved error reporting.
```

**Review by Frassle — APPROVED** [review]

**Comment by lukehoban** [comment]
bors merge

**Pull request merged**